### PR TITLE
add copy ability to execution environments

### PR DESCRIPTION
**Pull request by rebeccahhh**
##### SUMMARY

Adding copy functionality/api endpoints to execution environments and wrote a test for creating an EE
this will tie into #7062 
originally also added the pull field from #9179 to AWXKit in a couple places and started work on an RBAC issue, but I moved that work into PRs: #9333 and #9346

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

 - API

##### AWX VERSION

```
awx: 17.0.1
```

### CODE REVIEW

File: awx/main/tests/functional/test_execution_environments.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import pytest
+from unittest import mock
+from rest_framework.exceptions import PermissionDenied
+
+from awx.main.models import (ExecutionEnvironment, Organization)
+
+
+@pytest.mark.django_db
+def test_execution_environment_creation(execution_environment, organization):
+    execution_env = ExecutionEnvironment.objects.create(name='Hello Environment', image='', organization=organization, managed_by_tower=False, credential=None, pull='missing')
```
Comment (nixocio) [inline awx/main/tests/functional/test_execution_environments.py]: @rebeccahhh can one copy an EE `managed_by_tower`?

File: awx/api/views/__init__.py [hunk 2]
```diff
@@ -708,6 +708,12 @@ class ExecutionEnvironmentJobTemplateList(SubListAPIView):
     relationship = 'unifiedjobtemplates'
 
 
+class ExecutionEnvironmentCopy(CopyAPIView):
+
```
Comment (nixocio) [inline awx/api/views/__init__.py:714]: Are we updating `summary_fields` to add the copy field as part of the `user_capabilities`?

**Comment by shanemcd** [comment]
Can we rename this PR to be a bit more descriptive?

File: awxkit/awxkit/api/pages/execution_environments.py [hunk 3]
```diff
@@ -15,15 +15,15 @@
...
-class ExecutionEnvironment(HasCreate, base.Base):
+class ExecutionEnvironment(HasCreate, HasCopy, base.Base):
 
     dependencies = [Organization, Credential]
     NATURAL_KEY = ('name',)
 
-    # fields are name, image, organization, managed_by_tower, credential
-    def create(self, name='', image='quay.io/ansible/ansible-runner:devel', credential=None, **kwargs):
+    # fields are name, image, organization, managed_by_tower, credential, pull
+    def create(self, name='', image='quay.io/ansible/ansible-runner:devel', credential=None, pull='missing', **kwargs):
```
Comment (shanemcd) [inline awxkit/awxkit/api/pages/execution_environments.py]: We should almost always use `None` as a default kwarg value. See this for more info: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

I see that we were not doing that here for `image`, we should fix that too.

File: awxkit/awxkit/api/pages/execution_environments.py [hunk 4]
```diff
@@ -33,13 +33,13 @@ def create_payload(self, name='', organization=Organization, **kwargs):
         payload.ds = DSAdapter(self.__class__.__name__, self._dependency_store)
         return payload
 
-    def payload(self, name='', image=None, organization=None, credential=None, **kwargs):
+    def payload(self, name='', image=None, organization=None, credential=None, pull='missing', **kwargs):
```
Comment (shanemcd) [inline awxkit/awxkit/api/pages/execution_environments.py]: Same as my comment above, the default value here should be `None`.

File: awx/main/tests/functional/test_execution_environments.py (hunk 1 above)
Comment (rebeccahhh) [reply awx/main/tests/functional/test_execution_environments.py]: @nixocio I'm not sure, @jbradberry do you know?

Comment (jbradberry) [reply awx/main/tests/functional/test_execution_environments.py]: @nixocio there's no rbac logic that prevents it.  The expectation is that (some) global EEs are going to be the ones that are `managed_by_tower=True`, but there's nothing currently stopping users from creating global ones without it or Org-attached ones with it.  Anyway, for the purposes of rbac, my expectation is that if you can see it you can copy it, as long as you have sufficient privileges to do so in the current context (e.g. you have an Org Admin role).

... that raises an interesting question:  should an Org Admin be able to copy a global EE into their Org, and have that Org attachment happen automatically as part of the copy operation?

File: awx/api/views/__init__.py (hunk 2 above)
Comment (jbradberry) [reply awx/api/views/__init__.py:714]: Good call, we should do that.

File: awx/main/tests/functional/test_execution_environments.py (hunk 1 above)
Comment (nixocio) [reply awx/main/tests/functional/test_execution_environments.py]: Thanks for your explanation @jbradberry. Regarding the last question an Org Admin can be the admin of more than one Org, then automatically attaching the Org will not be a good fit, right? Right now for Org Admins to create an EE they must provide an Org.

File: awxkit/awxkit/api/pages/execution_environments.py (hunk 3 above)
Comment (jbradberry) [reply awxkit/awxkit/api/pages/execution_environments.py]: Python strings aren't mutable, and are perfectly fine as default function arguments.

Comment (rebeccahhh) [reply awxkit/awxkit/api/pages/execution_environments.py]: pulled out into #9333

File: awx/api/views/__init__.py (hunk 2 above)
Comment (rebeccahhh) [reply awx/api/views/__init__.py:714]: done!

**Review by jbradberry — APPROVED** [review]
This looks reasonable to me.

**Review by shanemcd — APPROVED** [review]
Amazing!

File: awxkit/awxkit/api/pages/execution_environments.py [hunk 5]
```diff
@@ -39,8 +39,6 @@ def payload(self, name='', image=None, organization=None, credential=None, pull=
             image=image or random_title(10),
             organization=organization.id if organization else None,
             credential=credential.id if credential else None,
-            pull=pull,
-            **kwargs
```
Comment (rebeccahhh) [inline awxkit/awxkit/api/pages/execution_environments.py]: I think these 2 lines must have been deleted when I was cherry-picking commits

**Comment by jbradberry** [comment]
Making sure to add this here, so it doesn't get missed -- we should allow for the copy of managed EEs, but copying such an EE should result in a new EE that has `managed_by_tower=False`.

**Review by shanemcd — APPROVED** [review]

**Pull request merged**